gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageFileSystem.java

```diff
@@ -414,22 +443,85 @@ public void delete(URI path, boolean recursive) throws IOException {
 
     List<FileInfo> bucketsToDelete = new ArrayList<>();
     (fileInfo.getItemInfo().isBucket() ? bucketsToDelete : itemsToDelete).add(fileInfo);
-
     coopLockOp.ifPresent(o -> o.persistAndScheduleRenewal(itemsToDelete, bucketsToDelete));
     try {
-      deleteInternal(itemsToDelete, bucketsToDelete);
-
+      deleteInternalWithFolders(itemsToDelete, listOfFolders, bucketsToDelete);
       coopLockOp.ifPresent(CoopLockOperationDelete::unlock);
     } finally {
       coopLockOp.ifPresent(CoopLockOperationDelete::cancelRenewal);
     }
-
     repairImplicitDirectory(parentInfoFuture);
   }
 
-  /** Deletes all items in the given path list followed by all bucket items. */
+  /**
+   * Return the bucket name if exists
+   *
+   * @param path
+   * @return bucket name
+   */
+  private String getBucketName(@Nonnull URI path) {
+    checkState(!Strings.isNullOrEmpty(path.getAuthority()), "Bucket name cannot be null");
+    return path.getAuthority();
+  }
+
+  /**
+   * Returns the folder name if exists else return empty string.
+   *
+   * @param path
+   * @return folder path
+   */
+  private String getFolderName(@Nonnull URI path) {
+    checkState(path.getPath().startsWith(PATH_DELIMITER), "Invalid folder name");
```

Comment:
include the path details along with the error message. Also at other places.

Reply:
Done